File: UX-Music-Wear/Shared/Models/Song.swift
import Foundation

/// Mirrors the Go scanner.Song struct (JSON schema must remain compatible)
struct Song: Codable, Identifiable, Hashable {
    let id: String
    let path: String
    let title: String
    let artist: String
    let album: String
    let albumArtist: String
    let year: Int
    let genre: String
    let duration: Double
    let trackNumber: Int
    let discNumber: Int
    let fileSize: Int64
    let fileType: String
    let sampleRate: Int?
    let bitDepth: Int?

    // artwork is excluded — fetched separately via /wear/artwork/{id}

    enum CodingKeys: String, CodingKey {
        case id, path, title, artist, album, year, genre, duration, fileSize, fileType
        case albumArtist = "albumartist"
        case trackNumber, discNumber, sampleRate, bitDepth
    }

    var displayTitle: String { title.isEmpty ? (path as NSString).lastPathComponent : title }
    var displayArtist: String { artist.isEmpty ? "Unknown Artist" : artist }
    var displayAlbum: String { album.isEmpty ? "Unknown Album" : album }

    var formattedDuration: String {
        let total = Int(duration)
        let minutes = total / 60
        let seconds = total % 60
        return String(format: "%d:%02d", minutes, seconds)
    }
}

File: UX-Music-Mobile-Legacy/lib/models/song.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'song.freezed.dart';
part 'song.g.dart';

/// Mirrors the Go scanner.Song struct (internal/scanner/scanner.go:14-31).
/// JSON keys match the Go json tags exactly.
@freezed
abstract class Song with _$Song {
  const factory Song({
    required String id,
    required String path,
    @Default('') String title,
    @Default('') String artist,
    @Default('') String album,
    @JsonKey(name: 'albumartist') @Default('') String albumArtist,
    @Default(0) int year,
    @Default('') String genre,
    @Default(0.0) double duration,
    @Default(0) int trackNumber,
    @Default(0) int discNumber,
    @Default(0) int fileSize,
    @Default('') String fileType,
    int? sampleRate,
    int? bitDepth,
    // SHA256 hash of "albumArtist---album" — used to construct artwork URLs.
    // Added by the /wear/songs endpoint; null on older server versions.
    @JsonKey(name: 'artworkId') @Default('') String artworkId,
  }) = _Song;

  const Song._();

  factory Song.fromJson(Map<String, dynamic> json) => _$SongFromJson(json);

  String get displayTitle => title.isEmpty ? 'Unknown Title' : title;
  String get displayArtist => artist.isEmpty ? 'Unknown Artist' : artist;
  String get displayAlbum => album.isEmpty ? 'Unknown Album' : album;

  String get formattedDuration {
    final minutes = duration ~/ 60;
    final seconds = (duration % 60).truncate();
    return '$minutes:${seconds.toString().padLeft(2, '0')}';
  }
}
